File: 백준/Silver/11659. 구간 합 구하기 4/구간 합 구하기 4.py
import sys

N,M = map(int,sys.stdin.readline().split())
arr = list(map(int,sys.stdin.readline().split()))
memo = [0] * (N+1)
memo[0] = arr[0]

for i in range(1,N):
    memo[i] = memo[i-1]+arr[i]
memo = [0] + memo

for i in range(M):
    a,b = map(int,sys.stdin.readline().split())
    print(memo[b]-memo[a-1])

File: 백준/Silver/11659. 구간 합 구하기 4/구간 합 구하기 4.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int[] arr = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        int[] result = new int[N];
        result[0] = arr[0];
        for(int i = 2; i<=N; i++){
            result[i-1] = result[i-2] + arr[i-1];
        }

        for(int i = 0; i< M ; i++){
            StringTokenizer check = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(check.nextToken());
            int b = Integer.parseInt(check.nextToken());
            System.out.println(result[b-1]- result[a-1] + arr[a-1]);
        }
    }
}
